cocotb/scheduler.py

```diff
@@ -598,15 +612,22 @@ def execute_external(func, _waiter):
                 self.log.debug("Execution of external routine done %s" % threading.current_thread())
             _waiter.thread_done()
 
-        waiter = external_waiter()
-        thread = threading.Thread(group=None, target=execute_external,
-                                  name=func.__name__ + "_thread",
-                                  args=([func, waiter]), kwargs={})
+        @coroutine
+        def wrapper():
+            waiter = external_waiter()
+            thread = threading.Thread(group=None, target=execute_external,
+                                      name=func.__name__ + "_thread",
+                                      args=([func, waiter]), kwargs={})
+
+            waiter.thread = thread
+            self._pending_threads.append(waiter)
+
+            yield waiter.event.wait()
 
-        waiter.thread = thread
-        self._pending_threads.append(waiter)
+            ret = ext.result  # raises if there was an exception
+            raise ReturnValue(ret)
 
-        return waiter
```

Comment:
Part C (dest) - note that this is just src, with `run_in_executor` replaced by the lines on the left